adr/0019-secrets.md

```diff
@@ -0,0 +1,63 @@
+
+## Decision Drivers <!-- optional -->
+
+- Secrets are a key feature of application platforms.
+- Up until wasmCloud 1.1, secrets had to be exposed in plaintext in order to work with wasmCloud applications.
+- Secrets were the most requested wasmCloud feature in the community Q2 roadmap session.
+
+## Considered Options
+
+- Leave security of secrets up to the security of NATS and TLS
+- Require secrets to be stored in a secrets backend and accessed via a capability provider at runtime
+- Support a specific secrets backend (e.g. Vault), and tie wasmCloud to that secrets backend.
+- **Support secrets as a first-class primitive in wasmCloud with a pluggable backend API**
+
+## Decision Outcome
+
+We've chosen to support secrets as a first-class primitive in wasmCloud with a pluggable backend API. This will enable wasmCloud to adhere to a very simple, pluggable, API, and support any and all secrets backends with a lightweight secrets-backend. This decision provides the best experience to users now and in the future, given that wasmCloud does not manage secrets directly and will not be a blocker for developers & companies that have their own secrets management solution.
+
+We've designed the `wasmcloud:secrets` interfaces specifically as a generic secrets interface that can be upstreamed in the future. It consists of two interfaces, `store` and `reveal`, which allow the design of interfaces that pass around a _resource_ handle to a secret rather than the value itself. Then, the component or provider that actually needs to use the secret can reveal and use the secret when needed. Interfaces that deal with sensitive values, personal data, passwords, etc can be specifically designed to never transmit the secret itself.
+
+wasmCloud requests secrets over NATS at a configured `--secrets-topic` at component start, provider start, and link creation time on behalf of the component/provider. Contained in the encrypted SecretRequest is the embedded JWT of the host, the component or provider, the application that component or provider belongs to, and any additional metadata fields that the backend needs in order to authorize the request. The request/response model to/from a secrets backend is done using XKeys that are generated on-the-fly, and no keys need to be centrally managed or shared between wasmCloud and the backends. Each SecretRequest uses its own XKey and thus cannot be replayed to a secrets backend. Secrets are then stored in-memory (for components) using the [secrecy](https://crates.io/crates/secrecy) crate, which zeroes out the memory the secret occupied as soon as it's dropped. Because all of these requests are done at start time, rather than when the secret is accessed in the component or provider, the component or provider will fail to start if the secret is not accessible. Additionally, if the secrets backend goes down or has a temporary outage, the operation of wasmCloud applications will not be interrupted.
```

Comment:
We should really be referring to topics as subjects universally.

Reply:
I only call them topics because in the wasi:messaging discussion that's the more standard language, but given that these are NATS subjects I agree and that should probs be what we use

Reply:
I only call them topics because in the wasi:messaging discussion that's the more standard language, but given that these are NATS subjects I agree and that should probs be what we use